### DISCUSSION TITLE
Executing scripts in non-interactive mode?

### DISCUSSION THREAD

**Discussion by tzakharko** [REditorSupport/vscode-R#1178] [General]
I am currently looking at VScode as a potential replacement for my current editor (Sublime). My R workflow usually involves invoking R scripts non-interactively (i.e.  via `Rscript script.r`), but from what I've gathered this extension always starts an interactive R console and executes a `source()` command into it. Is it possible to run R non-interactively and capture the (ANSI) output of the script or is it not supported?

**Comment by andycraig** [comment]
Hi, this extension does focus on interactive R use but you can use VS Code's built-in features to run code non-interactively. For example, if you want to run a file using Rscript you can open a terminal and run `Rscript script.r`.

VS Code also provides Tasks for automating flows that would normally be done with command-line scripts. You could create a Task that when run calls Rscript on the current file. This section of the Tasks documentation talks about using the current file: https://code.visualstudio.com/docs/editor/tasks#_variable-substitution

Does that sound like it would do the job? If not, can you provide more detail about how you use Rscript in Sublime? Then we can try to suggest an equivalent VS Code approach.

**Reply by tzakharko** [reply]
Thanks for your reply and the info! Essentially, I am looking for functionality to launch (potentially multiple) R scripts and display their output (including plots etc.).  I am often working on fairly complex data processing pipelines where the entire script hierarchy needs to be re-run on code changes, which doesn't work well with the the interactive sessions. Of course, there is always the option of invoking scripts manually but it's cumbersome. 

I have tried the approach with tasks, setting a default task to invoke `"Rscript ${file}`, which works, but with limitations. First, I can't seem to run multiple R scripts in this fashion  (it appears VScode links tasks to current folder/project, so when I try to run a different script it complains that the task is already running). And the other problem is that I can't display plots in the terminal, so I have to save them into external images and view them with separate tools. This is also problematic with Sublime, but at least there is a Terminus extension that can display images via ANSI escape codes. 

The editor that handles these workflows perfectly is TextMate 2, where the script output code is captured in a dynamic HTML document and displayed in a separate panel, including plots and rich output. But TM2 is very dated at this point and doesn't support other features that are common to modern editors. 

It would be great if the VScode plugin were to extend its functionality to support these workflows, maybe even doing something similar to TM2 with dynamically captured output (if VSCode even allows it). I would be willing to contribute implementing these features. What I am envisioning is two run script command flavours, one interactive and another non-interactive. But of course, it's not my project and it is entirely possible that the maintainers do not see room for this kind of functionality. It would certainly make the package more complicated.

**Reply by andycraig** [reply]
Thanks for this extra info!

It looks like the VS Code folks investigated image display in the terminal but didn't pursue it, so I don't think this is possible in VS Code at the moment. https://github.com/microsoft/vscode/issues/44436

A few options I can think of:
1. It might be possible to create a VS Code [webview](https://code.visualstudio.com/api/extension-guides/webview) that acts like a terminal but also displays images. I don't think it would need any connection with R (e.g., it would presumably work for Python too) so it could probably be its own extension. I haven't worked with webviews before so I don't have a sense of how hard this would be. 
2. Change the scripts to R Markdown files. You wouldn't get the output and images appearing as you ran them, but at the end you'd have HTML documents where you'd see all the output.
3. Since things need to be re-run whenever code changes, the {targets} package might allow an approximation of your current process but in an interactive session.

I think option 1 is what you want, as options 2 and 3 are basically asking you to change your workflow. But option 1 would probably also be the most work.
